ntoskrnl/ps/quota.c

```diff
@@ -18,123 +17,500 @@ EPROCESS_QUOTA_BLOCK PspDefaultQuotaBlock;
+ * quota spin lock has to be held when returning quotas.
+ */
+_Requires_lock_held_(PspQuotaLock)
+VOID
+NTAPI
+PspReturnQuotasOnDestroy(
+    _In_ PEPROCESS_QUOTA_BLOCK QuotaBlock)
+{
+    ULONG PsQuotaTypeIndex;
+    SIZE_T QuotaToReturn;
+
+    /*
+     * We must be in a dispatch level interrupt here
+     * as we should be under a spin lock.
+     */
+    ASSERT(KeGetCurrentIrql() == DISPATCH_LEVEL);
+
+    /* Make sure that the quota block is not plain garbage */
+    ASSERT(QuotaBlock);
+
+    /* Loop over the Process quota types */
+    for (PsQuotaTypeIndex = PsNonPagedPool; PsQuotaTypeIndex < PsPageFile; PsQuotaTypeIndex++)
+    {
+        /* The amount needed to return to Mm is the limit and return fields */
+        QuotaToReturn = QuotaBlock->QuotaEntry[PsQuotaTypeIndex].Limit + QuotaBlock->QuotaEntry[PsQuotaTypeIndex].Return;
+        MmReturnPoolQuota(PsQuotaTypeIndex, QuotaToReturn);
+    }
+}
+
+/**
+ * @brief
+ * Releases some of excess quotas in order to attempt
+ * free up some resources. This is done primarily in
+ * in case the Memory Manager fails to raise the quota
+ * limit.
+ *
+ * @param[in] QuotaType
+ * Process pool quota type.
+ *
+ * @param[out] ReturnedQuotas
+ * A pointer to the returned amount of quotas
+ * back to Memory Manager.
+ *
+ * @return
+ * Nothing.
+ *
+ * @remarks
+ * The function releases excess paged or non
+ * paged pool quotas. Page file quota type is
+ * not permitted. Furthermore, quota spin lock
+ * has to be held when returning quotas.
+ */
+_Requires_lock_held_(PspQuotaLock)
+VOID
+NTAPI
+PspReturnExcessQuotas(
+    _In_ PS_QUOTA_TYPE QuotaType,
+    _Outptr_ PSIZE_T ReturnedQuotas)
+{
+    PLIST_ENTRY PspQuotaList;
+    PEPROCESS_QUOTA_BLOCK QuotaBlockFromList;
+    SIZE_T AmountToReturn = 0;
+
+    /*
+     * We must be in a dispatch level interrupt here
+     * as we should be under a spin lock.
+     */
+    ASSERT(KeGetCurrentIrql() == DISPATCH_LEVEL);
+
+    /*
+     * Loop over the quota block lists and reap
+     * whatever quotas we haven't returned which
+     * is needed to free up resources.
+     */
+    for (PspQuotaList = PspQuotaBlockList.Flink;
+         PspQuotaList != &PspQuotaBlockList;
+         PspQuotaList = PspQuotaList->Flink)
+    {
+        /* Gather the quota block from the list */
+        QuotaBlockFromList = CONTAINING_RECORD(PspQuotaList, EPROCESS_QUOTA_BLOCK, QuotaList);
+
+        /*
+         * Gather any unreturned quotas and cache
+         * them to a variable.
+         */
+        AmountToReturn += InterlockedExchangeSizeT(&QuotaBlockFromList->QuotaEntry[QuotaType].Return, 0);
+
+        /*
+         * If no other process is taking use of this
+         * block, then we have to check that the limit
+         * is way over the usage. If that's the case,
+         * then we'll shrink the limit up on par with
+         * the usage and the additional amount to return
+         * to the Memory Manager is that limit minus the
+         * usage.
+         */
```

Comment:
Nit: this comment explains "what" right now. Perhaps add the "why": if ProcessCount=0, that means this block only has "shared" pool quota, which cannot increase (because there are no processes left to take it). Therefore the remaining quota isn't going to be used by anyone and we can safely take it.